src/compression.cpp

```diff
@@ -0,0 +1,88 @@
+#include <stdexcept>
+
+#include <lz4frame.h>
+
+#include "sparrow_ipc/compression.hpp"
+
+namespace sparrow_ipc
+{
+//     CompressionType to_compression_type(org::apache::arrow::flatbuf::CompressionType compression_type)
+//     {
+//         switch (compression_type)
+//         {
+//             case org::apache::arrow::flatbuf::CompressionType::LZ4_FRAME:
+//                 return CompressionType::LZ4;
+// //             case org::apache::arrow::flatbuf::CompressionType::ZSTD:
+// //                 // TODO: Add ZSTD support
+// //                 break;
+//             default:
+//                 return CompressionType::NONE;
+//         }
+//     }
+
+    std::vector<std::uint8_t> compress(const org::apache::arrow::flatbuf::CompressionType compression_type, std::span<const std::uint8_t> data)
+    {
+        if (data.empty())
+        {
+            return {};
+        }
+        switch (compression_type)
+        {
+            case org::apache::arrow::flatbuf::CompressionType::LZ4_FRAME:
+            {
+                const std::int64_t uncompressed_size = data.size();
+                const size_t max_compressed_size = LZ4F_compressFrameBound(uncompressed_size, nullptr);
+                std::vector<std::uint8_t> compressed_data(max_compressed_size);
+                const size_t compressed_size = LZ4F_compressFrame(compressed_data.data(), max_compressed_size, data.data(), uncompressed_size, nullptr);
+                if (LZ4F_isError(compressed_size))
+                {
+                    throw std::runtime_error("Failed to compress data with LZ4 frame format");
+                }
+                compressed_data.resize(compressed_size);
+                return compressed_data;
+            }
+            default:
+                return {data.begin(), data.end()};
+        }
+    }
+
+    std::vector<std::uint8_t> decompress(const org::apache::arrow::flatbuf::CompressionType compression_type, std::span<const std::uint8_t> data)
+    {
+        if (data.empty())
+        {
+            return {};
+        }
```

Comment:
Hmm, this should never happen according to the Flatbuffers spec. Did you encounter this situation somewhere?

Reply:
Hmm, the validity buffers being empty with compression is the only valid use case for now (because of the nulls optimization), and we check that exactly at the moment of getting that buffer, see https://github.com/QuantStack/sparrow-ipc/pull/30/commits/bc5415c0709ecb639af0e9542c1a4ceea2fde864.
Were you thinking of somewhere else?

Btw, I'm wondering about what's allowed and what's not, @pitrou :
If the compression is there and the data is empty (whatever the buffers), shouldn't we have at least the size header set to `-1`? (which doesn't seem to be the case for the empty validity buffers).

I added a `TODO` to check all this.

Reply:
> Does that mean that we actually shouldn't compress the validity buffers? Or anything else except from data buffers?

All buffers should be treated the same, i.e. undergo optional compression with a size header in front.

Reply:
Oh, that's interesting, thank you. That should definitely be called out in the Flatbuffers spec.

Reply:
Opened https://github.com/apache/arrow/issues/47918 for this.

Reply:
Thanks!

Reply:
> Were you thinking of somewhere else?

Yes, that could be inside the `utils::get_buffer` function (or in a thin function on top of get_buffer since there seems to be valid case wher ethe buffer is empty). The idea is to detect invalid input as soon as possible instead of deep down the deserialization process, and avoid "defensive" programming with a lot of checks everywhere that should actually be done upfront once.